README.md
# Ftl

Faster Than Light. Fog Terminal Language. For Terminal Launching.

## Description

`ftl` is an AWS (fog-based) application for managing cloud instances from the command line.
Easier than chef or the ec2 command line tools. Smarter too.

## Installation

Install is easy.

    $ gem install ftl # or sudo gem install ftl

## Usage

    Usage: ftl [<config-options>] <command> [<command-options>]
      commands: start, kill, list, edit, connect, servers, tags, images, snapshots, volumes
      examples:
        ftl launch ninja                   # Launches an instance named 'ninja'
        ftl list                           # Shows running instances and status
        ftl connect ninja                  # Connects to instance named 'ninja'
        ftl kill ninja                     # Kills instances matching name /ninja/
        ftl kill i-123456                  # Kills instance with id i-123456
        ftl spot ninja 0.02                # Request spot instance using template :ninja for $0.02
        ftl cancel sir-123456              # Cancel spot instance with id sir-123456
        ftl spots                          # Shows AWS spot requests
        ftl images                         # Shows AWS images
        ftl snapshots                      # Shows AWS snapshots
        ftl tags                           # Shows AWS tags
        ftl volumes                        # Shows AWS volumes
        ftl headers servers                # Show possible headers for servers
        ftl headers volumes                # Show possible headers for volumes
        ftl edit                           # Edit ftl.yml with your $EDITOR
        ftl --config=~/ftl.yml servers     # Uses custom config file
        ftl --headers=id,tags.Name servers # Uses specified headers
        ftl --version                      # Show version number

## Getting started

`ftl` depends on a configuration file. When you run it for the first time,
it will create a simple configuration file at `~/.ftl/ftl.yml`. Note that the starter file is incomplete;
you will need to supply at least your Amazon AWS security credentials. You can open the
configuration file in the editor named by the `$EDITOR` environment variable by running `ftl edit`.

### Configuration file

In addition to the AWS security credentials, the configuration file contains parameters for
launching instances and connecting to them.

    ACCESS_KEY_ID:      # Amazon AWS access key
    SECRET_ACCESS_KEY:  # Amazon AWS secret access key

    :keys:
      ninja-keypair:        ~/.ec2/id_rsa-ninja-keypair

    :templates:
      :ninja:
        :ami:               ami-a29943cb  # Ubuntu 12.04 LTS Precise Pangolin
        :username:          ubuntu
        :keypair:           ninja-keypair
        :instance_type:     m1.small
        :tags:              {}
        :groups:
          - default
          - apache
          - postgresql

#### The :templates section

`:templates` is a hash of server names and the EC2 parameters used to launch them.

* :ami -- the ID of the AMI to use
* :availability_zone -- the availability zone into which the instance should be launched
* :groups -- a list of names and/or IDs of security groups to assign to the instance
* :instance_type -- the name of the EC2 instance size (e.g. 'm1.small')
* :ip_private -- the private IP address to assign to the instance (optional)
* :ip_public -- the public IP address to assign to the instance (optional)
* :keypair -- the name of the keypair installed in the instance when launched; this is important
              for connecting to the instance once it has been launched
* :subnet_id -- the ID of the subnet into which the instance should be launched (optional)
* :tags -- a hash of tags to be set on the instance; 'Name' will be automatically set from the
           server name
* :username -- the username to use when connecting to the instance

#### A note on security groups

Security groups defined in a virtual private cloud (VPC) can be specified only by ID.

#### The :keys section

`:keys`, if defined, is a hash of keypair names and the names of files on the local filesystem
containing them. If there is an entry in `:keys` with the same name as the server, the file will be
passed in the `-i` option to `ssh` when connecting.

### Launching an instance

Once a server has been configured, it can be launched by name.

    ftl launch ninja

To connect to it, run `ftl connect ninja`.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

© 2012 Matthew Petty 

* Twitter: [@lodestone](http://twitter.com/lodestone) 
* Github: [@lodestone](http://github.com/lodestone) 
* Email: [lodestone@gmail.com](mailto:lodestone@gmail.com)

Changelog
# Release Version 0.3.3 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Feature: Add ftl address - returns a publicly accessible ip or hostname for a named instance.

# Release Version 0.3.2 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Feature: Add ftl templates (BETA)
* Feature: Load yml files from templates subdirectory if available (BETA)
* Bug Fix: ssh command usernames

# Release Version 0.3.1 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Allow spot requests to accept --count parameter

# Release Version 0.3.0 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Feature: Add ftl <script> to execute arbitrary ruby locally within ftl environment
* Feature: Add ftl ips to show Elastic IP addresses
* Feature: Add ftl stop to stop running instances
* Feature: Add ftl start to start running instances
* Feature: Add option to pass in spot request instance count
* Feature: Add ftl terminate to specifically terminate any instance vs only running ones
* Bug fix: Spot requests were launching as regular instances 
* Bug fix: Other minor bug fixes and cleanups


# Release Version 0.2.4 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Add ftl <action> <server> to execute arbitrary commands on running instances
* Add ftl sample_config to see the latest sample ftl.yml
* Add ftl images # Shows AMIs associated with your account

# Release Version 0.2.3 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  
* Add ftl status
* find_instances returns exact match preferentially
* Update README with more examples and setup
* Adds :keys config parameter

# Release Version 0.2.2 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Adds templates feature to config file
* Start and cancel spot requests
* Execution without parameters shows help screen
* Fixes various bugs

# Release Version 0.2.0 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* First revamped release handling AWS in a more generic way

Previous releases were focused on starting up Pair Programming servers. 
The latest releases can also be used for this but are more useful in a generic 
AWS situation.

test/helper.rb
require 'rubygems'
require 'test/unit'
require 'shoulda'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'ftl'

class Test::Unit::TestCase
end


test/test_ftl.rb
require 'helper'

class TestFtl < Test::Unit::TestCase
  should "probably rename this file and start testing for real" do
    flunk "hey buddy, you should probably rename this file and start testing for real"
  end
end


lib/ftl/server.rb
require 'sinatra'
require 'active_support'
require 'right_aws'
require 'sdb/active_sdb'
require 'json'
require 'open-uri'

# require 'fog'
# ENV["FOG_RC"] = Will need this if we switch to Fog

class PairingMachine < Aws::ActiveSdb::Base
end

module FTL

  class Server < Sinatra::Base

    ACCESS_KEY_ID     = ENV['ACCESS_KEY_ID']     || YAML.load_file(ENV['HOME']+'/.ftl/ftl.yml')['ACCESS_KEY_ID']
    SECRET_ACCESS_KEY = ENV['SECRET_ACCESS_KEY'] || YAML.load_file(ENV['HOME']+'/.ftl/ftl.yml')['SECRET_ACCESS_KEY']
    INSTANCE_SCRIPT   = ENV['INSTANCE_SCRIPT']   || YAML.load_file(ENV['HOME']+'/.ftl/ftl.yml')['INSTANCE_SCRIPT']

    before do
      Aws::ActiveSdb.establish_connection(ACCESS_KEY_ID, SECRET_ACCESS_KEY)
      PairingMachine.create_domain # Raises no error if it already exists, but blows up if it doesn't
    end

    get '/machines' do
      # List the machines
      pms = PairingMachine.find(:all)
      ec2 = Aws::Ec2.new(ACCESS_KEY_ID, SECRET_ACCESS_KEY)
      pms.each do |pm|
        i = ec2.describe_instances(pm.attributes['aws_instance_id'])[0]
        pm[:dns_name] = i[:dns_name]
        pm.save
      end
      pms.collect{|pm| pm.attributes }.to_json
    end

    post '/machines' do
      # Create new machine
      return "Need to name this machine" unless params[:name]
      return "Please provide an ami" unless params[:ami]
      pm = PairingMachine.new(params)
      pm[:instance_type] ||= 'c1.medium'
      ec2 = Aws::Ec2.new(ACCESS_KEY_ID, SECRET_ACCESS_KEY)
      if INSTANCE_SCRIPT.nil?
        instance_script = '#!' 
      else
        instance_script = INSTANCE_SCRIPT[/^#!/] ? INSTANCE_SCRIPT : URI.parse(INSTANCE_SCRIPT).read
      end
      i = ec2.launch_instances(params[:ami], :user_data => instance_script, :instance_type => pm[:instance_type])[0]
      pm[:aws_instance_id] = i[:aws_instance_id]
      pm[:dns_name] = i[:dns_name]
      pm.save
      pm.attributes.to_json
    end

    delete '/machines' do
      # Destroy the machine by short name
      pms = PairingMachine.find(:all)
      pms = pms.select{|pm| pm.attributes['name'].first[params['name']] }
      return "{message: \"Provide the name of the machine(s) to be terminated\"}" if params['name'].nil? || pms.blank?
      ec2 = Aws::Ec2.new(ACCESS_KEY_ID, SECRET_ACCESS_KEY)
      instances = pms.collect{|pm| pm.attributes['aws_instance_id'].first }
      ec2.terminate_instances(instances)
      pms.each {|pm|
        pm.delete 
      }
      "{message: \"Shutdown machines matching '#{params['name']}'\"}"
    end

  end

end



lib/ftl/client.rb
module Ftl

  def self.help_message
    %Q{
    Usage: [bold]ftl[/] \[<config-options>\] <command> \[<command-options>\]
      commands: start, kill, list, edit, connect, servers, tags, images, snapshots, volumes
      examples:
        ftl launch ninja                   # Launches an instance named 'ninja'
        ftl list                           # Shows running instances and status
        ftl connect ninja                  # Connects to instance named 'ninja'
        ftl kill ninja                     # Kills instances matching name /ninja/
        ftl kill i-123456                  # Kills instance with id i-123456
        ftl spot ninja 0.02                # Request spot instance using template :ninja for $0.02
        ftl cancel sir-123456              # Cancel spot instance with id sir-123456
        ftl spots                          # Shows AWS spot requests
        ftl images                         # Shows AWS images
        ftl snapshots                      # Shows AWS snapshots
        ftl tags                           # Shows AWS tags
        ftl volumes                        # Shows AWS volumes
        ftl headers servers                # Show possible headers for servers
        ftl headers volumes                # Show possible headers for volumes
        ftl edit                           # Edit ftl.yml with your $EDITOR
        ftl images                         # Show AMIs associated with your account
        ftl <action> ninja                 # Execute the <action> (specified in ftl.yml) on the remote server instance
        ftl <script> ninja                 # Execute the <script> (specified in ftl.yml) on your local machine
        ftl --config=~/ftl.yml servers     # Uses custom config file
        ftl --headers=id,tags.Name servers # Uses specified headers
        ftl --version                      # Show version number
    }
  end

  def self.help(*args)
    Formatador.display_line help_message
  end

  class Client

    SINGLE_COMMANDS = %w{ edit sample }

    attr_reader :con
    attr_accessor :options

    def initialize(args=nil, opts={})
      if args && args.length > 0
        arg = args.reverse.pop
        @args = [arg, args - [arg]]
        load_config(opts)
        if (!SINGLE_COMMANDS.include?(arg))
          @con = Fog::Compute.new(:provider => 'AWS', :aws_secret_access_key => options['SECRET_ACCESS_KEY'], :aws_access_key_id => options['ACCESS_KEY_ID'])
        end
        send(*@args)
      else
        Ftl.help
      end
    end

    def launch_instance(args)
      display "Spinning up FTL..."
      opts = options
      opts = options.merge(options[:templates][args.first.to_sym]) if !options[:templates][args.first.to_sym].nil?

      opts[:group_ids] = (opts[:group_ids] || []) + opts[:groups].select { | group | group.to_s =~ /^sg-[0-9a-f]{8}$/ }
      opts[:groups] = opts[:groups].reject { | group | group.to_s =~ /^sg-[0-9a-f]{8}$/ }

      launcher = options.delete(:launcher) || :servers
      server = con.send(launcher).create(
                                  :user_data          => opts[:user_data],
                                  :key_name           => opts[:keypair],
                                  :groups             => opts[:groups],
                                  :security_group_ids => opts[:group_ids],
                                  :image_id           => opts[:ami],
                                  :availability_zone  => opts[:availability_zone],
                                  :flavor_id          => opts[:instance_type],
                                  :username           => opts[:username],
                                  :tags               => opts[:tags].merge(:Name => args.first),
                                  :subnet_id          => opts[:subnet_id],
                                  :private_ip_address => opts[:ip_private],
                                  :ip_address         => opts[:ip_address],
                                  :price              => opts[:price],
                                  :instance_count     => opts[:count]
                                 )

      display server
      display "Executing :post_script..." if opts[:post_script]
      eval(opts[:post_script]) if opts[:post_script]
    end

    def launch(args={})
      guard(args.first, "Please provide a short name for instance\n\t[bold]ftl[/] launch <name>")
      options.merge(options[:templates][args.first.to_sym]) if !options[:templates][args.first.to_sym].nil?
      server = launch_instance(args)
    end
    alias :up     :launch
    alias :spinup :launch
    alias :create :launch
    alias :new    :launch

    def spot(args={})
      guard(args[0], "Please provide a short name for instance\n\t[bold]ftl[/] spot <name> <price>")
      guard(args[1], "Please provide a price for spot request\n\t[bold]ftl[/] spot <name> <price>")
      display "Spinning up FTL..."
      options.merge(options[:templates][args.first.to_sym]) if !options[:templates][args.first.to_sym].nil?
      options[:price] = args[1] || options[:templates][args.first.to_sym][:price] || options[:price]
      options.merge!(:launcher => :spot_requests)
      server = launch_instance(args)
    end
    alias :request :spot

    def spots(args={})
      con.spot_requests.table(_headers_for(:spot_requests))
    end
    alias :spot_requests :spots

    def cancel(args={})
      guard(args.first, "Please provide the id for the spot request to cancel.") 
      spot = con.spot_requests.get(args.first)
      if spot && spot.destroy
        display "Canceled Spot Instance #{spot.id}."
      else
        display "Whups, spot instance not found!"
      end
    end

    def running_instances(name)
      find_instances(name).select{|i| i.state == "running" }
    end

    def running_instance(name)
      instances = running_instances(name)
      instances.first if instances
    end

    def connect(args={})
      if server = running_instance(args.first)
        puts(ssh_command(server))
        exec(ssh_command(server))
      else
        display "Typo alert! No server found!"
      end
    end
    alias :x :connect
    alias :ssh :connect

    def status(args={})
      guard(args, :message => "Please provide the name/id of a server (ftl status <server>)")
      server = find_instance(args.first)
      display server
    end
    alias :st :status

    def start(args={})
      display "Bringing \"#{args.first}\" server back to life."
      find_instance(args.first).start
    end
    alias :run :start

    def stop(args={})
      display "Stopping \"#{args.first}\" server."
      find_instance(args.first).stop
    end
    alias :pause :stop

    def destroy(args={})
      guard(on_what, "Please provide the name (or partial name for the instance(s) you want to delete. For instance, like: ftl destroy ninja")
      display "Spinning down FTL..."
      instances = find_instances(on_what).select{|i| i.state == 'running' }
      if !instances.empty?
        instances.map(&:destroy)
        display "Destroyed [bold]\[#{instances.map(&:id).join(', ')}\][/]"
      else
        display "No instances found"
      end
    end
    alias :d         :destroy
    alias :delete    :destroy
    alias :kill      :destroy
    alias :down      :destroy
    alias :shutdown  :destroy

    def terminate(args={})
      guard(on_what, "Please provide the name (or partial name for the instance(s) you want to terminate. For instance, like: ftl destroy ninja")
      display "Spinning down FTL..."
      instances = find_instances(on_what)
      if !instances.empty?
        instances.map(&:destroy)
        display "Destroyed [bold]\[#{instances.map(&:id).join(', ')}\][/]"
      else
        display "No instances found"
      end
    end
    alias :t :terminate

    def info(args={})
      display find_instance(args.first)
    end
    alias :i :info

    def list(opts)
      opts = [:servers] if opts.empty?
      con.send(opts.first).table(_headers_for(opts.first))
    end
    alias :l  :list
    alias :ls :list

    def image(args={})
      Formatador.display_table(con.images.find(:id => args.first))
    end
    alias :img :image

    # TODO: Make this better by including more details of block devices
    def images(args={})
      hashes = con.describe_images('Owner' => 'self').body['imagesSet'].collect do |hash|
        h = {}
        h[:image_id]  = hash['imageId']
        h[:name]  = hash['name']
        h[:rootDeviceType] = hash['rootDeviceType']
        h
      end 
      puts Formatador.display_table(hashes)
    end
    alias :describe_images :images

    def ip_addresses
      con.addresses
    end

    def ips(args={})
      # TODO add server "name"
      addrs = con.describe_addresses.body['addressesSet'] #.collect do |addr|
        # addr
      # end 
      Formatador.display_table addrs
    end
    alias :addresses :ips

    def headers(args={})
      what = args.first || :servers
      display "Showing header options for #{what}"
      display con.send(what).first.attributes.keys
    end

    def server_instances(args={})
      @servers ||= (con.servers.all||[])
    end

    def edit(args={})
      display "You need to set [bold]$EDITOR[/] environment variable" if ENV['EDITOR'].nil?
      `$EDITOR ~/.ftl/ftl.yml`
    end

    def sample(args={})
      puts File.open(File.dirname(__FILE__) + "/../resources/ftl.yml").read
    end

    def command
      @args[0]
    end

    def secondary_arguments
      @args[1] 
    end

    def on_what 
      secondary_arguments.first
    end

    def volume(args={})
      arg = args.is_a?(String) ? args : on_what
      display vol = con.volumes.select{|v| v.tags['Name'] == arg || v.id == arg }.first
      vol
    end

    def execute(args={})
      arg = args.is_a?(String) ? args : args[1]
      server = find_instances(on_what).select{|i| i.state == 'running' }.first
      puts %|#{ssh_command(server)} #{arg}|
      system(%|#{ssh_command(server)} #{arg}|)
    end
    alias :ex :execute

    def address(args={})
      if server = running_instance(args.first)
        puts server_address(server)
      else
        display "Typo alert! No server found!"
      end
    end
    alias :a :address

    ###########################################################################
    ## private                                                              ###
    ###########################################################################
    private
 
    def server_address(server)
      server[:public_ip_address] || server[:dns_name] || server[:private_ip_address]
    end

    def ssh_command(server)
      opt_key = " -i #{options[:keys][server[:key_name]]}" unless (options[:keys].nil? || options[:keys][server[:key_name]].nil?)
      hostname = server_address(server)
      if server.tags['Name']
        server_name =  server.tags['Name'].to_sym
        user_name = options[:templates][server_name][:username] 
      end
      user_name = 'root' if user_name.nil? || user_name.length == 0
      "ssh#{opt_key} #{user_name}@#{hostname}"
    end

    def guard(arg, options={:message => "Please refer to ftl help"})
      if arg.nil?
        display options[:message]
        exit
      end
    end
 
    def ftl_yml
      File.open("lib/resources/ftl.yml").read
    end

    def load_config(opts={})
      # TODO So ugly, make this less shitty. Such a common pattern.
      if opts[:config]
        default_config_file = opts[:config]
      else
        default_config_name = 'ftl.yml'
        default_config_dir  = '/.ftl/'
        default_config_home = "#{ENV['HOME']}#{default_config_dir}"
        default_config_file = "#{default_config_home}#{default_config_name}"
        if Dir.exist?(default_config_home) # Directory exists
          if !File.exist?(default_config_file) # File does not
            File.open(default_config_file, 'w') {|f| f << ftl_yml }
          end
        else # Directory does not exist
          Dir.mkdir(default_config_home)
          File.open(default_config_file, 'w') {|f| f << ftl_yml }
        end
      end
      @options = YAML.load_file(default_config_file)
      @options = @options.merge(opts)
      load_extended_config_yaml
      puts "======Please open #{default_config_file} and set ACCESS_KEY_ID and SECRET_ACCESS_KEY====\n\n" if aws_credentials_absent?
    end

    def yaml_config(*args)
      puts @options
    end

    def templates(args)
      if args.first
        p @options[:templates][args.first.to_sym]
      else
        @options[:templates].each do |template|
          puts template.first
        end
      end
    end

    # TODO Make :scripts and :actions the same way
    def load_extended_config_yaml
      Dir.glob("#{ENV['HOME']}/.ftl/templates/*.yml").each do |file|
        yaml = YAML.load_file(file)
        name = File.basename(file).gsub(".yml", '')
        @options[:templates][name.to_sym]=@options[:templates][:defaults]
        @options[:templates][name.to_sym].merge!(yaml)
      end
    end

    def find_instance(name)
      find_instances(name).first
    end

    def find_instances(name)
      id_match  = server_instances.select{|i| i[:id] == name } if name[/^i-/]
      exact_tag_match = server_instances.select{|i| !i.tags.nil? && i.tags['Name'] && i.tags['Name'] == name }
      tag_match = server_instances.select{|i| !i.tags.nil? && i.tags['Name'] && i.tags['Name'][name] }
      id_match || exact_tag_match || tag_match
    end

    def _headers_for(object)
      return options[:headers].map(&:to_sym) if options[:headers]
      case object
      when :snapshots
        [:id, :volume_id, :state, :volume_size, :description, :"tags.Name"]
      when :spot_requests
        [:id, :image_id, :availability_zone, :price, :flavor_id, :state, :request_type, :launched, :created_at]
      when :volumes
        [:server_id, :id, :size, :snapshot_id, :availability_zone, :state, :"tags.Name"]
      when nil
        [:id, :image_id, :flavor_id, :availability_zone, :state, :created_at, :"tags.Name"]
      else
        [:id, :image_id, :flavor_id, :availability_zone, :state, :created_at, :"tags.Name"]
      end
    end

    def aws_credentials_absent?
      options['ACCESS_KEY_ID'].nil? || options['SECRET_ACCESS_KEY'].nil?
    end

    def display(message)
      msg = message.is_a?(String) ? message : message.inspect
      Formatador.display_line(msg)
    end

    def eval_action(script, args)
      # TODO Complete the script to handle multiple servers, like:
      # ftl bundle server1 server2 server3
      # server = find_instance(on_what)
      server = find_instances(on_what).select{|i| i.state == 'running' }.first
      eval(script, binding)
    end

    def eval_script(script, args)
      eval(script, binding)
    end

    def local(cmd)
      puts output = %x|#{cmd}|
      output
    end

    # def remote(cmd)
    #   %x|#{ssh_command(server)} #{cmd}|
    # end

    def method_missing(*args)
      method = args.first.to_sym
      if con.respond_to? method
        results = con.send(method)
        display results.table(_headers_for(method))
      elsif options[:actions][method]
        eval_action(options[:actions][method], args)
      elsif options[:scripts][method]
        eval_script(options[:scripts][method], args)
      else
        Ftl.help
      end
    end

  end

end
